### PR TITLE
21 http extended method

### PR DESCRIPTION
**Pull request by miniupnp**
better have this in a PR for discussing.

1) Tests are missing.

2) I'm still not convinced by the wording.
Why not call the function "ext_methods_cb" or "ext_methods_func"
That's not a compare only callback.

### CODE REVIEW

**Comment by miniupnp** [comment]
I can add tests based on those I done with my old implementation ( 001edb5ac881c4c988a8331dfc522edef63f3c92 )
But I'd like to make sure this naming stuff is ok. @azat any idea ?

**Comment by miniupnp** [comment]
tests : 8f586d0b0f69e4f7b013b0cca17d9a7eaf37f98f

**Comment by miniupnp** [comment]
ping ?

**Comment by ellzey** [comment]
I thought we discussed this. Does https://github.com/libevent/libevent/commit/35487dc5828b07aea514f5161b6d4fa23a7c998en not work for you?

**Comment by miniupnp** [comment]
Well, I'm globally fine with 35487dc5828b07aea514f5161b6d4fa23a7c998e
I fixed it with 7b520746f60a1338c7692d86ff94828689a00455 and
added/changed comments / documentation in 7485bbe2b9228d8e965d65d63dcc1a326c1ddd1a

I may have missed it, but I've not read your answer to my question : 

> Why not call the the function "extended_method_func" ?
> ```
evhttp_set_extended_method_func()
> evhttp_connection_set_extended_method_func()
```

https://github.com/libevent/libevent/pull/327#commitcomment-16322083

**Comment by miniupnp** [comment]
If you prefer to keep the current name, I guess you can just merge the PR (and the test : https://github.com/miniupnp/libevent/tree/21_http_extended_method-tests )

**Comment by ellzey** [comment]
OK, that looks good. We're about to make a 2.1 release, will revisit and merge afterwards.

**Comment by kiler129** [comment]
I've applied changes onto current master, you can use following patch: https://gist.github.com/kiler129/9502c558a3268c30602f913898a495ce

Additionally I've prepared simple echo-server utilizing new capabilities:
```C
#include <sys/types.h>
#include <sys/time.h>
#include <sys/queue.h>
#include <stdlib.h>
#include <err.h>
#include <event2/event.h>
#include <event2/event_compat.h>
#include <event2/http.h>
#include <event2/http_compat.h>
#include <event2/buffer.h>
#include <string.h>

#define EVHTTP_REQ_CUSTOM ((EVHTTP_REQ_MAX) << 1)
static int ext_method_cb(struct evhttp_extended_method * p)
{
	if(p == NULL) return -1;
	if(p->method) {
		if(strcmp(p->method, "CUSTOM") == 0) {
			p->type = EVHTTP_REQ_CUSTOM;
			p->flags = 0;	/*EVHTTP_METHOD_HAS_BODY*/
			return 0;
		}
	} else {
		if(p->type == EVHTTP_REQ_CUSTOM) {
			p->method = "CUSTOM";
			return 0;
		}
	}
	return -1;
}

static void
http_custom_cb(struct evhttp_request *req, void *arg)
{
	struct evbuffer *evb = evbuffer_new();
	evbuffer_add_printf(evb, "Hello CUSTOM!");
	evhttp_send_reply(req, HTTP_OK, "CUSTOM OK", evb);
	evbuffer_free(evb);
}

int main(int argc, char **argv)
{
    struct evhttp *httpd;
    event_init();
    httpd = evhttp_start("127.0.0.1", 8080);

    evhttp_set_extended_method_cmp(httpd, ext_method_cb);
    evhttp_set_allowed_methods(httpd, EVHTTP_REQ_CUSTOM);

    evhttp_set_cb(httpd, "/", http_custom_cb, NULL);

    event_dispatch();
    evhttp_free(httpd);

    return 0;
}
```

Since I'm not an C guy it took ma a while to actually get it to compile, so I'm kindly leaving how-to for googlers:
0. Compile libevent as usual
1. `gcc -I/home/greg/libevent-customhttp/libevent-master/include http-echo.c -L/home/greg/libevent-customhttp/libevent-master/.libs -levent`
2. `LD_LIBRARY_PATH=../.libs ./a.out`
3. Play with it:

screenshot 2016-11-27 01 38 13

**Comment by miniupnp** [comment]
You want me to rebase on current master ?

**Comment by kiler129** [comment]
Looking at the [milestone for 2.1](https://github.com/libevent/libevent/milestone/1) I personally think not much is gonna change, so it would be great if you can rebase for current master.

**Comment by azat** [comment]
It's been awhile, so does anybody still interested in this (@miniupnp @NathanFrench)?
If so, can you rebase against current master and add unit tests?

**Comment by azat** [comment]
Skip the "add unit tests" part, I found the [test] (https://github.com/libevent/libevent/pull/327#issuecomment-200804893) @miniupnp but why it does not included into this PR?

**Comment by miniupnp** [comment]
OK, I'm trying to rebase on master.

**Comment by miniupnp** [comment]
I've rebased. Maybe the test should be rewritten...

File: http.c [hunk 1]
```diff
@@ -1941,6 +1988,25 @@ evhttp_parse_request_line(struct evhttp_request *req, char *line, size_t len)
...
+		 * returns a 0 value.
+		 */
+		struct evhttp_extended_method ext_method;
+
+		ext_method.method = method;
+		ext_method.type = EVHTTP_REQ_UNKNOWN_;
+
+		if (req->evcon->ext_method_cmp &&
+			    req->evcon->ext_method_cmp(&ext_method) == 0) {
+		    /* TODO: make sure the other fields in ext_method are
```
Comment (azat) [inline http.c]: I guess we can safely ignore this, since we do not use them later

File: http.c [hunk 2]
```diff
@@ -353,10 +360,49 @@ evhttp_method(enum evhttp_cmd_type type)
...
+		 *       flags.
+		 */
+		ext_method.method = NULL;
+		ext_method.type   = type;
+		ext_method.flags  = tmp_flags;
+
+		if (evcon->ext_method_cmp != NULL &&
+			evcon->ext_method_cmp(&ext_method) == 0) {
+
+		    if (ext_method.type != type) {
```
Comment (azat) [inline http.c]: In this case, at least `method` should be set to `NULL` for now

File: http.c [hunk 3]
```diff
@@ -2476,7 +2525,9 @@ evhttp_connection_base_bufferevent_new(struct event_base *base, struct evdns_bas
 {
 	struct evhttp_connection *evcon = NULL;
 
-	event_debug(("Attempting connection to %s:%d\n", address, port));
+	event_debug(("%s %s:%d\n",
+	             (dnsbase?"New connection from":"Attempting connection to"),
+	             address, port));
```
[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: My initial commit was : https://github.com/miniupnp/libevent/commit/8de03d1a18aa8f9068f75da950ce99368f67e267
It is to have different message when being the HTTP server or the HTTP client.

[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: I think at some point it was squashed in @errzey commit... I don't remember why/how, that's 3 years ago :(

File: http.c [hunk 4]
```diff
@@ -299,9 +301,12 @@ evhttp_htmlescape(const char *html)
...
+evhttp_method_(struct evhttp_connection *evcon,
+              enum evhttp_cmd_type type, ev_uint16_t *flags)
 {
-	const char *method;
+	struct evhttp_extended_method ext_method;
+	const char *method    = NULL;
+	ev_uint16_t tmp_flags = EVHTTP_METHOD_HAS_BODY;
 
 	switch (type) {
 	case EVHTTP_REQ_GET:
```
[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: I may be missing something here.
see `http_method_may_have_body`

File: test/regress_http.c [hunk 5]
```diff
@@ -71,13 +71,35 @@ static struct event_base *exit_base;
 
 static char const BASIC_REQUEST_BODY[] = "This is funny";
 
+/* extended Method */
```
[the remark this answers is not in the document]
Comment (miniupnp) [reply test/regress_http.c]: done

File: http.c [hunk 6]
```diff
@@ -353,10 +360,49 @@ evhttp_method(enum evhttp_cmd_type type)
...
+			/* XXX: when this etuns, and the method is NULL,
+			 * the caller of this function just blindly adds
+			 * it to the evbuffer! We need to figure out a
+			 * way to propigate that error all the way down
+			 * the chain.
+			 *
+			 * For now we just break, but it really needs
+			 * to return NULL.
+			 */
+			event_debug(("%s: method_cmp modified type from %u to %u, not allowd",
```
[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: callback was better

File: include/event2/http.h [hunk 7]
```diff
@@ -552,6 +571,24 @@ enum evhttp_cmd_type {
 	EVHTTP_REQ_MOVE    = 1 << 15,
 };
 
+#define EVHTTP_REQ_MAX EVHTTP_REQ_PATCH
```
[the remark this answers is not in the document]
Comment (miniupnp) [reply include/event2/http.h]: yep bd9ed4a57a8beddb7fe7d9a7a0bb1d7a8f172f28
 see #796 also.

File: include/event2/http.h [hunk 8]
```diff
@@ -571,7 +571,7 @@ enum evhttp_cmd_type {
 	EVHTTP_REQ_MOVE    = 1 << 15,
 };
 
-#define EVHTTP_REQ_MAX EVHTTP_REQ_PATCH
+#define EVHTTP_REQ_MAX EVHTTP_REQ_MOVE
```
Comment (azat) [inline include/event2/http.h]: This should be fixed in a separate patch (the original `EVHTTP_REQ_UNKNOWN_` const)

As for exporting this in a public header this has pros and cons:
- pros: can be used in user code
- cons: will be used in user code and hence can break programs after new methods will be added (although I guess that users will stick to the last method in this define anyway, so not sure which one I prefer)

Comment (miniupnp) [reply include/event2/http.h]: I don't see how it is related to  `EVHTTP_REQ_UNKNOWN_` ?
user should use values > `EVHTTP_REQ_MAX ` and < `EVHTTP_REQ_UNKNOWN_`

File: http.c [hunk 9]
```diff
@@ -4041,6 +4098,12 @@ evhttp_set_allowed_methods(struct evhttp* http, ev_uint16_t methods)
 	http->allowed_methods = methods;
 }
 
+void
+evhttp_set_extended_method_cmp(struct evhttp* http,
+    int (*cmp)(struct evhttp_extended_method *)) {
```
[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: for the function type ???
`evhttp_set_cb` has no typdeF...

File: http.c [hunk 10]
```diff
@@ -465,8 +513,8 @@ evhttp_make_header_request(struct evhttp_connection *evcon,
 	    method, req->uri, req->major, req->minor);
 
 	/* Add the content length on a post or put request if missing */
-	if ((req->type == EVHTTP_REQ_POST || req->type == EVHTTP_REQ_PUT) &&
-	    evhttp_find_header(req->output_headers, "Content-Length") == NULL){
+	if ((flags & EVHTTP_METHOD_HAS_BODY) &&
+	    evhttp_find_header(req->output_headers, "Content-Length") == NULL) {
```
Comment (miniupnp) [inline http.c]: hum this patch may break something as more methods are defined to "have a body" because they could (but sometimes shoudn't)

Comment (miniupnp) [reply http.c]: maybe we could have flags EVHTTP_METHOD_MAY_HAS_BODY and EVHTTP_METHOD_MUST_HAS_BODY instead of just EVHTTP_METHOD_HAS_BODY

File: include/event2/http.h (hunk 8 above)
Comment (azat) [reply include/event2/http.h]: Sorry, my mistake, didn't look carefully

File: http.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (azat) [reply http.c]: >for the function type ???

yep

>evhttp_set_cb has no typdeF...

that is not great

File: http.c (hunk 10 above)
Comment (azat) [reply http.c]: >hum this patch may break something as more methods are defined to "have a body" because they could (but sometimes shoudn't)

Indeed

>maybe we could have flags EVHTTP_METHOD_MAY_HAS_BODY and EVHTTP_METHOD_MUST_HAS_BODY instead of just EVHTTP_METHOD_HAS_BODY

 I didn't find anything in [rfc2616](https://tools.ietf.org/html/rfc2616) that POST "MUST" have body, hence I don't like this naming.

And this logic looks indeed strange, I would replace it with `if (evhttp_method_may_have_body(req->type) && !evhttp_find_header(req->output_headers, "Content-Length") && evbuffer_get_length(req->output_buffer)) ...` (but requires some testing)

File: http.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (azat) [reply http.c]: Yeah, the birth of this patch is old enough, but thanks for getting back to it!
So this hunk should be dropped

File: http.c [hunk 11]
```diff
@@ -2576,6 +2628,12 @@ int evhttp_connection_set_flags(struct evhttp_connection *evcon,
 	return 0;
 }
 
+void
+evhttp_connection_set_extended_method_cmp(struct evhttp_connection *evcon,
+	int (*cmp)(struct evhttp_extended_method *)) {
```
Comment (azat) [inline http.c]: I would prefer to see typedef for this

File: http.c [hunk 12]
```diff
@@ -4036,11 +4094,17 @@ evhttp_set_default_content_type(struct evhttp *http,
 }
 
 void
-evhttp_set_allowed_methods(struct evhttp* http, ev_uint16_t methods)
+evhttp_set_allowed_methods(struct evhttp* http, ev_uint32_t methods)
```
Comment (azat) [inline http.c]: This requires separate patch

File: test/regress_http.c [hunk 13]
```diff
@@ -71,13 +71,37 @@ static struct event_base *exit_base;
...
+/* defines an extended HTTP method "CUSTOM"
+ * without body */
+#define EVHTTP_REQ_CUSTOM	((EVHTTP_REQ_MAX) << 1)
+
+static int ext_method_cb(struct evhttp_extended_method *p)
+{
+	if (p == NULL)
+		return -1;
+	if (p->method) {
+		if(strcmp(p->method, "CUSTOM") == 0) {
```
Comment (azat) [inline test/regress_http.c]: style: missing whitespace and please check the whole patchset.

(we have `checkpatch.sh` and `.clang-format` in the git root, but they can be out dated)

Comment (miniupnp) [reply test/regress_http.c]: hum clang-format-diff is not available on my machine :(

File: http.c (hunk 12 above)
Comment (miniupnp) [reply http.c]: isn't 0757534f6e6eed1e37588ada9a38c661999651dc "separate" enough ?
I think you can cherry-pick it in master

**Comment by azat** [comment]
> isn't 0757534 "separate" enough ?

Fair enough, sorry to bother.

**Comment by miniupnp** [comment]
> > isn't 0757534 "separate" enough ?
>
> Fair enough, sorry to bother.
No pb, do you want me to open a specific Pull request for it ?

**Comment by azat** [comment]
> No pb, do you want me to open a specific Pull request for it ?

Nope, this one is enough (either it will be merged with this PR or
applied separately)

**Comment by miniupnp** [comment]
I rebased. I'm not sure I fixed all the issues.

File: http.c (hunk 10 above)
Comment (miniupnp) [reply http.c]: I rewrote the test in a way it won't break anything (forcing Content-Length with PUT and POST)

File: http.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (azat) [reply http.c]: I don't see how `dnsbase == "connection from"`, I can call `evhttp_connection_base_bufferevent_new(NULL, evdns_base_new(), ...)`

File: http.c [hunk 14]
```diff
@@ -353,10 +360,49 @@ evhttp_method(enum evhttp_cmd_type type)
...
+		ext_method.type   = type;
+		ext_method.flags  = tmp_flags;
+
+		if (evcon->ext_method_cmp != NULL &&
+			evcon->ext_method_cmp(&ext_method) == 0) {
+
+		    if (ext_method.type != type) {
+			/* XXX: when this etuns, and the method is NULL,
+			 * the caller of this function just blindly adds
+			 * it to the evbuffer! We need to figure out a
```
[the remark this answers is not in the document]
Comment (azat) [reply http.c]: ping

File: http.c (hunk 11 above)
Comment (azat) [reply http.c]: I don't see typedef here, you mixed things a little bit, since you added typedef for `evhttp_set_cb` (while this patch is not about it), but skipped `evhttp_connection_set_extended_method_cmp`

File: include/event2/http.h [hunk 15]
```diff
@@ -552,6 +571,24 @@ enum evhttp_cmd_type {
...
+
+/** structure to allow users to define their own HTTP methods
+ * @see evhttp_set_extended_method_cmp
+ * @see evhttp_connection_set_extended_method_cmp */
+
+/**
+ * @brief stucture that is passed to (and modified by) the
+ * extended method callback function
+ */
+struct evhttp_extended_method {
```
[the remark this answers is not in the document]
Comment (azat) [reply include/event2/http.h]: ping

File: http.c (hunk 11 above)
Comment (azat) [reply http.c]: Oh, you just forgot convert this helper too (like `evhttp_set_extended_method_cmp`)

File: test/regress_http.c (hunk 13 above)
Comment (azat) [reply test/regress_http.c]: ping

File: http.c (hunk 10 above)
Comment (azat) [reply http.c]: >forcing Content-Length with PUT and POST

I would vote to avoid forcing Content-Length for PUT/POST, since it is valid to send POST without Content-Length

Comment (miniupnp) [reply http.c]: well this is a feature dating back to a3bb4a035
https://github.com/libevent/libevent/commit/a3bb4a035fe27dcdf52158d115c4caea153dabf5#diff-f6b6aab933915430c8e4f4a5dfa99eb2R201
I don't think we should change it. Or then in a specific commit

**Comment by azat** [comment]
> Or then in a specific commit

Agree

File: http.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: well I'm not sure what you are asking. I think that comment was written by @errzey

File: http.c (hunk 2 above)
Comment (miniupnp) [reply http.c]: I'm returning NULL now

File: http.c (hunk 1 above)
Comment (miniupnp) [reply http.c]: I don't understand what you mean

File: http.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: When I wrote this patch 3 years ago it fixed the bug.
I'm open for a better way to detect whether the http connection is a server or a client one

File: http.c (hunk 11 above)
Comment (miniupnp) [reply http.c]: let me check

File: include/event2/http.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (miniupnp) [reply include/event2/http.h]: so you want `struct evhttp_extended_method` to be renamed to `struct evhttp_ext_method`. What else ?

File: test/regress_http.c (hunk 13 above)
Comment (miniupnp) [reply test/regress_http.c]: As said above, `checkpatch.sh` doesn't work on my machine, because` clang-format-diff ` is not available.
Maybe you can run it and send me the results by email.

File: http.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: I guess the `event_debug()` line was written when only HTTP client was implemented. HTTP server was later implemented but the debug message was not updated.

By the way, you're right it looks like the test doesn't work anymore.
I think it would be better to remove the log from the `evhttp_connection_base_bufferevent_new()` function and set it in the calling functions.

File: http.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (azat) [reply http.c]: I think that this comment can be removed, also take a lot at the alignment around this comment, it uses tabs *and* spaces, should be tabs only

File: http.c (hunk 10 above)
Comment (azat) [reply http.c]: Agree

File: http.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (azat) [reply http.c]: >When I wrote this patch 3 years ago it fixed the bug.

This hunk cannot fix any bug.
So I think that this should be removed.

>I'm open for a better way to detect whether the http connection is a server or a client one

Do you really use debug logging in libevent? I guess not.
And even if you do, this is a separate issue, totally unrelated to this PR

[the remark this answers is not in the document]
Comment (miniupnp) [reply http.c]: yes it is unrelated to this PR.
But yes I used debug logging in libevent while I was adding custom HTTP request support.

[the remark this answers is not in the document]
Comment (azat) [reply http.c]: >I think it would be better to remove the log from the evhttp_connection_base_bufferevent_new() function and set it in the calling functions.

First of all, like I wrote in the previous comment, let's keep this PR clean to the unrelated changes.

But to make it clear, evhttp_connection_base_bufferevent_new() is an exported function, so if you will move the logging message into the calling function it will be lost for the users that calls it directly (and it is pretty common).

File: test/regress_http.c (hunk 13 above)
Comment (azat) [reply test/regress_http.c]: It can use `uncrustify` too.

I cannot do this for you right now, maybe later, but I would suggest you to check this patches manually (or writing your own script) if you cannot use `checkpatch.sh` (actually it requires some modifications since `clang-format-diff` is now `clang-format-diff.py` and has different options style, and also it reports `ENOENT` error to me).

*P.S. I don't think it is a great idea to ask other people to check coding style issues in your patches*

Comment (miniupnp) [reply test/regress_http.c]: I'm sorry. most of the patch if from @errzey anyway... 

I've run `Uncrustify-0.68_f` on `http.c`, it makes me a lot of differences. Some of them are a bit strange.
```
@@ -4923,7 +4942,8 @@
                case PART_FRAGMENT:
                        cp += strlen(cp);
                        break;
-               };
+               }
+               ;
                return cp;
        }
```
I'll check all changes that are related to this PR and correct.

File: http.c (hunk 11 above)
Comment (miniupnp) [reply http.c]: should be ok now

File: test/regress_http.c (hunk 13 above)
Comment (miniupnp) [reply test/regress_http.c]:
```
$ git show 29ce63b19d0cf1e0ca07ed3ca3fed1fc9e89039a | ./checkpatch.sh -U -p
Not implemented
$ ./checkpatch.sh -U -r 29ce63b19d0cf1e0ca07ed3ca3fed1fc9e89039a
$ ./checkpatch.sh -U -d http.c
[... patch ...]
```

Comment (azat) [reply test/regress_http.c]: >I've run Uncrustify-0.68_f on http.c, it makes me a lot of differences

It is a bad idea to run uncrustify on the whole file, since the coding style in it vary

>Some of them are a bit strange.

Indeed, so basically just run it only on you patch

>$ git show 29ce63b19d0cf1e0ca07ed3ca3fed1fc9e89039a | ./checkpatch.sh -U -p
>Not implemented

But I guess it is not possible to run it only on the patch right now, egh.

So I would suggest you to look through the patch manually, this should take too much time, since the patch is pretty small

File: http.c (hunk 11 above)
Comment (azat) [reply http.c]: Looks good, thanks!

File: test/regress_http.c (hunk 13 above)
Comment (miniupnp) [reply test/regress_http.c]: I have a problem. you said : 

> style:  `const char *foo()` over `const char * foo()` (i.e. for functions the same as for variable definitions, e.t.c.)

but uncrustify outputs 
```
-static const char *evhttp_method_(struct evhttp_connection *evcon,
-       enum evhttp_cmd_type type, ev_uint16_t *flags);
+static const char * evhttp_method_(struct evhttp_connection *evcon,
+                                  enum evhttp_cmd_type type, ev_uint16_t *flags);
```
I think it is because of this : 
https://github.com/libevent/libevent/blob/master/.uncrustify#L15
```
sp_after_ptr_star                        = remove
sp_after_ptr_star_func                   = add
```

should I fix `.uncustify` ?

Comment (azat) [reply test/regress_http.c]: The basic rule is: use coding style like in the code around

>should I fix .uncustify ?

Maybe it is a good idea, but I need to check, since AFAIR I created this config to get minimal changes to the current sources

**Comment by miniupnp** [comment]
I reviewed the whole patch for coding style issues with the help of uncrustify. I hope I missed none.
I also added a few lines in whatsnew-2.2.txt

File: include/event2/http.h [hunk 16]
```diff
@@ -270,6 +270,7 @@ EVENT2_EXPORT_SYMBOL
 void evhttp_set_ext_method_cmp(struct evhttp *http,
 	evhttp_ext_method_cb *cmp);
 
+typedef void evhttp_cb(struct evhttp_request *, void *);
```
Comment (azat) [inline include/event2/http.h]: First of all it overlaps with `evhttp_cb` in `http-internal.h` which is not a gread idea (although it requires `struct`)
Secondly it is used in other places.

So I'm going to fix this by myself, objections?

File: http-internal.h [hunk 17]
```diff
@@ -32,7 +32,7 @@ struct addrinfo;
 struct evhttp_request;
 
 /* Indicates an unknown request method. */
-#define EVHTTP_REQ_UNKNOWN_ (1<<15)
+#define EVHTTP_REQ_UNKNOWN_ (1<<31)
```
Comment (azat) [inline http-internal.h]: Replaced with just 0 in 799053d, rebased patch is 847f6f944a7a1aecff8a99d85ca05d819ead1b54

**Comment by azat** [comment]
>I reviewed the whole patch for coding style issues with the help of uncrustify. I hope I missed none.

Thanks! You fixed most of them.

>I also added a few lines in whatsnew-2.2.txt

This file is append-only, so I moved your section to the end of file and also make it a little bit more verbose/cleaner -- c7921dd3e286bf050124480c8b170ac533f54585

So the whole rebased patch set you will find here:
  https://github.com/azat/libevent.git 21_http_extended_method

Objections?

(I'm going to do some final cleanups other then this everything looks good)

**Comment by miniupnp** [comment]
>     I also added a few lines in whatsnew-2.2.txt
>
> This file is append-only, so I moved your section to the end of file
> and also make it a little bit more verbose/cleaner -- c7921dd
> <https://github.com/libevent/libevent/commit/c7921dd3e286bf050124480c8b170ac533f54585>
>
> So the whole rebased patch set you will find here:
> https://github.com/azat/libevent.git 21_http_extended_method
>
> Objections?
>
> (I'm going to do some final cleanups other then this everything looks
> good)
>
sorry for whatsnew-2.2.txt I though the top ... where the place to put
new stuff. I should have looked at the previous commits

I have no objection to your 21_http_extended_method branch !

File: include/event2/http.h (hunk 16 above)
Comment (miniupnp) [reply include/event2/http.h]: no objection

**Comment by azat** [comment]
So I squashed the commits and make you the primary author (with a note that this patch based on changes from @errzey)

**Comment by azat** [comment]
Applied, thanks for rebasing! (and sorry for such en enormous delay..)
